### DISCUSSION TITLE
Should the order of 'on conditions in a Decider affect the outcome of a Job?

### DISCUSSION THREAD

**Discussion by rojyates** [spring-projects/spring-batch#4686] [Q&A] [answered]
I have a decider that checks for the presence of a file. It has two possible outcomes defined as constants:

- NO_FILE
- FIE_EXISTS

**The outcome of the Job when the file is present unexpectedly depends on the order in which I specify the .on conditions in this Flow:**

I originally wrote the check with the NO_FILE check last:
**Option 1. 'fail condition checked last':**
```java
               .next(fileExistsDecider)
                .on(FILE_EXISTS).to(continueFlow)
                .on(NO_FILE).fail()
                .build();
```
When this is executed with no file (result NO_FILE), it behaves as expected and the job fails.
When the file is present, however, the job continues to execute as expected with all steps finishing as 'COMPLETED', but the overall Job finishes with 'FAILED'.

**Reversing the order of these .on conditions, changes the behaviour of the Job.**

**Option 2. 'success condition checked last':**
```java
               .next(fileExistsDecider)
                .on(NO_FILE).fail()
                .on(FILE_EXISTS).to(continueFlow)
                .build();
```
Now we have the fail condition checked first, and the success condition last.
When this is executed with no file (result NO_FILE), it behaves as expected and the job fails.
When the file is present, the job continues to execute as expected with all steps finishing as 'COMPLETED', and the overall job finishes with 'COMPLETED', as expected.

I would expect the order of these checks not to impact the final result of the job.
I've added a comment with this info into https://github.com/spring-projects/spring-batch/issues/4478 but thought to get some opinions...:

Is this a bug, or am I missing something?

**Comment by fmbenhassine** [comment, marked as answer]
> Should the order of 'on conditions in a Decider affect the outcome of a Job?

No, the order should not matter. This is a bug in Spring Batch, I added a comment on #4478: https://github.com/spring-projects/spring-batch/issues/4478#issuecomment-2430636958
